WorkloadManagementSystem/Agent/SiteDirector.py

```diff
@@ -1216,67 +1226,28 @@ def _writePilotScript(self, workingDirectory, pilotOptions,
 
   def updatePilotStatus(self):
-    """ Update status of pilots in transient states
+    """ Update status of pilots in transient and final states
     """
-    for queue in self.queueDict:
-      ce = self.queueDict[queue]['CE']
-      ceName = self.queueDict[queue]['CEName']
-      queueName = self.queueDict[queue]['QueueName']
-      ceType = self.queueDict[queue]['CEType']
-      siteName = self.queueDict[queue]['Site']
-
-      result = pilotAgentsDB.selectPilots({'DestinationSite': ceName,
-                                           'Queue': queueName,
-                                           'GridType': ceType,
-                                           'GridSite': siteName,
-                                           'Status': TRANSIENT_PILOT_STATUS,
-                                           'OwnerDN': self.pilotDN,
-                                           'OwnerGroup': self.pilotGroup})
-      if not result['OK']:
-        self.log.error('Failed to select pilots", ": %s' % result['Message'])
-        continue
-      pilotRefs = result['Value']
-      if not pilotRefs:
-        continue
-
-      result = pilotAgentsDB.getPilotInfo(pilotRefs)
-      if not result['OK']:
-        self.log.error('Failed to get pilots info from DB', result['Message'])
-        continue
-      pilotDict = result['Value']
-
-      stampedPilotRefs = []
-      for pRef in pilotDict:
-        if pilotDict[pRef]['PilotStamp']:
-          stampedPilotRefs.append(pRef + ":::" + pilotDict[pRef]['PilotStamp'])
-        else:
-          stampedPilotRefs = list(pilotRefs)
-          break
 
-      # This proxy is used for checking the pilot status and renewals
-      # We really need at least a few hours otherwise the renewed
-      # proxy may expire before we check again...
-      result = ce.isProxyValid(3 * 3600)
-      if not result['OK']:
-        result = gProxyManager.getPilotProxyFromDIRACGroup(self.pilotDN, self.pilotGroup, 23400)
-        if not result['OK']:
+    # Generate a proxy before feeding the threads to renew the ones of the CEs to perform actions
+    result = gProxyManager.getPilotProxyFromDIRACGroup(self.pilotDN, self.pilotGroup, 23400)
+    if not result['OK']:
+      return result
+    proxy = result['Value']
+
+    # Getting the status of pilots in a queue implies the use of remote CEs and may lead to network latency
+    # Threads aim at overcoming such issues and thus 1 thread per queue is created to
+    # update the status of pilots in transient states
+    with ThreadPoolExecutor(max_workers=len(self.queueDict)) as executor:
+      threads = []
+      for queue in self.queueDict:
+        threads.append(executor.submit(self._updatePilotStatusPerQueue, queue, proxy))
+
+      # as_completed yields completed threads and then we get their results to make sure no error occured
+      for future in as_completed(threads):
+        result = future.result()
```

Comment:
Do I understand correctly that this is the output from `_updatePilotStatusPerQueue`? That function always returns `None` doesn't it?

Reply:
Indeed, in this case the result is actually not checked. But if even some particular thread fails, the cycle should still continue and the subsequent code from this method should also execute. So, the behavior is OK but the coding is not good.

Reply:
@andresailer Yes you're right, thanks.
I removed the lines to get the results as it is not needed.
The cycle continues even if a thread fails and the coding should be better now.